File: libraries/dotnet/UrcConverter/Sources/Bms/Convert.fs
namespace UrcConverter.Sources.Bms

module Convert =

    open FsToolkit.ErrorHandling
    open UrcConverter
    open UrcConverter.Sources
    open UrcConverter.Sources.Bms.Model
    open Channels

    [<Literal>]
    let private MeasureUs = 240000000.0

    [<RequireQualifiedAccess>]
    type private EntryKind =
        | Bpm of float
        | Meter of int
        | Stop of float
        | Scroll of float
        | Object of int
        | Anchor

    type private ScanAcc =
        {
            Entries: (float * int * EntryKind) list
            Objects: (float * string * string) list
            Used: Set<int * char>
        }

    type private TimingScanState =
        {
            TimeUs: float
            PrevY: float
            PendingStop: float
            CurrentBpm: float option
            CurrentBeats: int
            BpmPoints: (int * float * int * int) list
            SvPoints: (int * float) list
            Anchors: int list
            Timed: Map<int, float>
        }

    let convertBms (chart: BmsChart) : Result<Chart, UrcError> =
        result {
            let! bpmInitial =
                match chart.Bpm with
                | Some bpm when bpm > 0.0 -> Ok bpm
                | _ -> Error(UrcError.Syntax(1, "missing or non-positive #BPM"))

            let maxMeasure =
                chart.Measures
                |> Map.keys
                |> Seq.fold max -1

            let boundaries =
                [| 0 .. maxMeasure |]
                |> Array.scan (fun acc m ->
                    let rate = Map.tryFind m chart.Rates |> Option.defaultValue 1.0
                    acc + rate) 0.0

            let initialAcc =
                {
                    Entries = [ (0.0, 0, EntryKind.Bpm bpmInitial) ]
                    Objects = []
                    Used = Set.empty
                }

            let rec scanMeasures (acc: ScanAcc) (m: int) : Result<ScanAcc, UrcError> =
                if m > maxMeasure then
                    Ok { acc with Objects = List.rev acc.Objects }
                else
                    let rate = Map.tryFind m chart.Rates |> Option.defaultValue 1.0
                    let prevRate = Map.tryFind (m - 1) chart.Rates |> Option.defaultValue 1.0
                    let meterEntries =
                        if rate <> prevRate then
                            let beats = rate * 4.0
                            if abs (beats - round beats) < 1e-9 && round beats >= 1.0 then
                                [ (boundaries[m], 3, EntryKind.Meter (int (round beats))) ]
                            else []
                        else []

                    let accWithMeter = { acc with Entries = meterEntries @ acc.Entries }

                    match Map.tryFind m chart.Measures with
                    | None -> scanMeasures accWithMeter (m + 1)
                    | Some measureMap ->
                        let rec scanChannels (cAcc: ScanAcc) (channels: (string * string list) list) : Result<ScanAcc, UrcError> =
                            match channels with
                            | [] -> Ok cAcc
                            | (channel, ids) :: restChannels ->
                                let rec scanIds (iAcc: ScanAcc) (idx: int) : Result<ScanAcc, UrcError> =
                                    if idx >= ids.Length then
                                        Ok iAcc
                                    else
                                        let obj = ids[idx]
                                        let y = boundaries[m] + (float idx / float ids.Length) * rate
                                        if Set.contains channel systemChannels then
                                            if obj <> "00" then
                                                match channel with
                                                | "03" ->
                                                    let digits = idValue obj chart.Base
                                                    let bpmVal = float ((digits / 36) * 16 + (digits % 36))
                                                    scanIds { iAcc with Entries = (y, 0, EntryKind.Bpm bpmVal) :: iAcc.Entries } (idx + 1)
                                                | "08" ->
                                                    match Map.tryFind obj chart.BpmDefs with
                                                    | Some bpmVal -> scanIds { iAcc with Entries = (y, 0, EntryKind.Bpm bpmVal) :: iAcc.Entries } (idx + 1)
                                                    | None -> Error(UrcError.Syntax(1, $"undefined #BPM{obj}"))
                                                | "09" ->
                                                    match Map.tryFind obj chart.StopDefs with
                                                    | Some stopVal -> scanIds { iAcc with Entries = (y, 1, EntryKind.Stop stopVal) :: iAcc.Entries } (idx + 1)
                                                    | None -> Error(UrcError.Syntax(1, $"undefined #STOP{obj}"))
                                                | _ ->
                                                    match Map.tryFind obj chart.ScrollDefs with
                                                    | Some scrollVal -> scanIds { iAcc with Entries = (y, 2, EntryKind.Scroll scrollVal) :: iAcc.Entries } (idx + 1)
                                                    | None -> Error(UrcError.Syntax(1, $"undefined #SCROLL{obj}"))
                                            else
                                                scanIds iAcc (idx + 1)
                                        else
                                            match channelKind channel with
                                            | None -> scanIds iAcc (idx + 1)
                                            | Some kind ->
                                                let used =
                                                    if obj <> "00" then Set.add (sideOf channel[0], channel[1]) iAcc.Used
                                                    else iAcc.Used
                                                if not (obj = "00" && kind <> "ln") then
                                                    let objIdx = iAcc.Objects.Length
                                                    let objects = (y, channel, obj) :: iAcc.Objects
                                                    let entries = (y, 4, EntryKind.Object objIdx) :: iAcc.Entries
                                                    scanIds { Entries = entries; Objects = objects; Used = used } (idx + 1)
                                                else
                                                    scanIds { iAcc with Used = used } (idx + 1)

                                scanIds cAcc 0
                                |> Result.bind (fun nextAcc -> scanChannels nextAcc restChannels)

                        scanChannels accWithMeter (Map.toList measureMap)
                        |> Result.bind (fun nextAcc -> scanMeasures nextAcc (m + 1))

            let! scanned = scanMeasures initialAcc 0
            let mode = detectMode chart.Pms scanned.Used

            let sortedEntries =
                (boundaries
                 |> Seq.map (fun y -> y, 5, EntryKind.Anchor)
                 |> List.ofSeq)
                @ scanned.Entries
                |> List.sortBy (fun (y, order, _) -> y, order)

            let groupedEntries =
                sortedEntries
                |> List.groupBy (fun (y, _, _) -> y)

            let initialTiming =
                {
                    TimeUs = 0.0
                    PrevY = 0.0
                    PendingStop = 0.0
                    CurrentBpm = None
                    CurrentBeats = 4
                    BpmPoints = []
                    SvPoints = []
                    Anchors = []
                    Timed = Map.empty
                }

            let foldTimingGroup (state: TimingScanState) (y: float, group: (float * int * EntryKind) list) =
                let timeUs =
                    match state.CurrentBpm with
                    | Some b -> state.TimeUs + (MeasureUs * (y - state.PrevY) / b)
                    | None -> state.TimeUs
                    + state.PendingStop

                let rec foldGroupEvents nextBpm nextBeats pendingStop scroll anchors timed (events: (float * int * EntryKind) list) =
                    match events with
                    | [] -> nextBpm, nextBeats, pendingStop, scroll, anchors, timed
                    | (_, _, kind) :: rest ->
                        match kind with
                        | EntryKind.Bpm v -> foldGroupEvents (Some v) nextBeats pendingStop scroll anchors timed rest
                        | EntryKind.Meter v -> foldGroupEvents nextBpm v pendingStop scroll anchors timed rest
                        | EntryKind.Stop v ->
                            let stopTime =
                                match nextBpm with
                                | Some b -> MeasureUs * v / b
                                | None -> 0.0
                            foldGroupEvents nextBpm nextBeats stopTime scroll anchors timed rest
                        | EntryKind.Scroll v -> foldGroupEvents nextBpm nextBeats pendingStop (Some v) anchors timed rest
                        | EntryKind.Object idx -> foldGroupEvents nextBpm nextBeats pendingStop scroll anchors (Map.add idx timeUs timed) rest
                        | EntryKind.Anchor -> foldGroupEvents nextBpm nextBeats pendingStop scroll (Shared.roundMs (timeUs / 1000.0) :: anchors) timed rest

                let nextBpm, nextBeats, pendingStop, scroll, nextAnchors, nextTimed =
                    foldGroupEvents state.CurrentBpm state.CurrentBeats 0.0 None state.Anchors state.Timed group

                let bpmPoints =
                    match nextBpm with
                    | Some bpm when nextBpm <> state.CurrentBpm || nextBeats <> state.CurrentBeats ->
                        (Shared.roundMs (timeUs / 1000.0), bpm, nextBeats, 4) :: state.BpmPoints
                    | _ ->
                        state.BpmPoints

                let svPoints =
                    match scroll with
                    | Some s -> (Shared.roundMs (timeUs / 1000.0), s) :: state.SvPoints
                    | None -> state.SvPoints

                {
                    TimeUs = timeUs
                    PrevY = y
                    PendingStop = pendingStop
                    CurrentBpm = nextBpm
                    CurrentBeats = nextBeats
                    BpmPoints = bpmPoints
                    SvPoints = svPoints
                    Anchors = nextAnchors
                    Timed = nextTimed
                }

            let timingState = (initialTiming, groupedEntries) ||> List.fold foldTimingGroup

            let! rawNotes = buildNotes chart mode scanned.Objects timingState.Timed
            let firstNoteTime =
                rawNotes
                |> List.filter (fun (_, _, t) -> t <> NoteType.LE)
                |> List.map (fun (t, _, _) -> t)
                |> function
                    | [] -> 0
                    | xs -> List.min xs

            let anchorMs = timingState.Anchors |> List.rev |> List.tryFind (fun time -> time >= firstNoteTime)

            let! timing =
                Shared.buildTiming (List.rev timingState.BpmPoints) (List.rev timingState.SvPoints) firstNoteTime ".bms" anchorMs

            let typeOrder = function
                | NoteType.N -> 0
                | NoteType.LS -> 1
                | NoteType.LE -> 2
                | NoteType.M -> 3
                | NoteType.F -> 4

            let urcNotes =
                rawNotes
                |> List.map (fun (t, lane, nt) ->
                    {
                        TimestampMs = t - firstNoteTime
                        Lane = lane
                        Type = nt
                    })
                |> List.sortBy (fun n -> n.TimestampMs, n.Lane, typeOrder n.Type)

            do! Shared.checkHoldOverlap urcNotes

            let keys, specialKeys, specialLanes = resolveLayout mode

            return
                {
                    FormatVersion = { Major = 1; Minor = 1 }
                    Metadata =
                        {
                            Original = if chart.Pms then "PMS" else "BMS"
                            Title = chart.Title |> Option.defaultValue "Unknown"
                            Artist = chart.Artist |> Option.defaultValue "Unknown"
                            Creator = "Unknown"
                            Version = chart.PlayLevel |> Option.defaultValue "Unknown"
                        }
                    Judgment = None
                    Layout =
                        {
                            Keys = keys
                            SpecialKeys = specialKeys
                            SpecialLanes = specialLanes
                        }
                    TimingPoints = timing
                    Notes = urcNotes
                }
        }

File: libraries/python/src/urc_converter/sources/bms/convert.py
"""Mapper from the BMS-family source model onto a URC chart."""

from itertools import groupby

from ...error import UrcError
from ...model import Chart, Layout, Metadata, Note, NoteType, Version
from .._shared import build_timing, check_hold_overlap, round_ms
from .model import BmsChart

_MEASURE_US = 240000000.0
_TYPE_ORDER = {NoteType.N: 0, NoteType.LS: 1, NoteType.LE: 2, NoteType.M: 3}

_SIDE_OF = {"1": 0, "5": 0, "D": 0, "2": 1, "6": 1, "E": 1}
_SYSTEM_CHANNELS = ("02", "03", "08", "09", "SC")
_LAYOUTS = {
	"5K": (5, 1, [0]),
	"7K": (7, 1, [0]),
	"10K": (10, 2, [0, 6]),
	"14K": (14, 2, [0, 8]),
	"PMS9": (9, 0, None),
	"PMS18": (18, 0, None),
}


def convert_bms(chart: BmsChart) -> Chart:
	"""Map a BMS-family chart onto a URC chart."""
	if chart.bpm is None or chart.bpm <= 0:
		raise UrcError("syntax", 1, "missing or non-positive #BPM")

	max_measure = max(chart.measures, default=-1)
	boundaries = [0.0]
	for measure in range(max_measure + 1):
		boundaries.append(boundaries[-1] + chart.rates.get(measure, 1.0))

	entries: list[tuple[float, int, str, float | int]] = [(0.0, 0, "bpm", chart.bpm)]
	objects: list[tuple[float, str, str]] = []
	used: set[tuple[int, str]] = set()

	for measure in range(max_measure + 1):
		rate = chart.rates.get(measure, 1.0)
		if rate != chart.rates.get(measure - 1, 1.0):
			beats = rate * 4.0
			if abs(beats - round(beats)) < 1e-9 and round(beats) >= 1:
				entries.append((boundaries[measure], 3, "meter", round(beats)))

		for channel, ids in chart.measures.get(measure, {}).items():
			for index, obj in enumerate(ids):
				y = boundaries[measure] + (index / len(ids)) * rate

				if channel in _SYSTEM_CHANNELS:
					if obj == "00":
						continue
					if channel == "03":
						digits = _id_value(obj, chart.base)
						entries.append((y, 0, "bpm", float((digits // 36) * 16 + digits % 36)))
					elif channel == "08":
						if obj not in chart.bpm_defs:
							raise UrcError("syntax", 1, f"undefined #BPM{obj}")
						entries.append((y, 0, "bpm", chart.bpm_defs[obj]))
					elif channel == "09":
						if obj not in chart.stop_defs:
							raise UrcError("syntax", 1, f"undefined #STOP{obj}")
						entries.append((y, 1, "stop", chart.stop_defs[obj]))
					else:
						if obj not in chart.scroll_defs:
							raise UrcError("syntax", 1, f"undefined #SCROLL{obj}")
						entries.append((y, 2, "scroll", chart.scroll_defs[obj]))
					continue

				kind = _channel_kind(channel)
				if kind is None:
					continue
				if obj != "00":
					used.add((_SIDE_OF[channel[0]], channel[1]))
				if obj == "00" and kind != "ln":
					continue
				objects.append((y, channel, obj))
				entries.append((y, 4, "object", len(objects) - 1))

	entries.extend((y, 5, "anchor", 0) for y in boundaries)

	mode = _detect_mode(chart.pms, used)

	bpm: float | None = None
	beats = 4
	time_us = 0.0
	prev_y = 0.0
	pending_stop = 0.0
	timed = [0.0] * len(objects)
	bpm_points: list[tuple[int, float, int, int]] = []
	sv_points: list[tuple[int, float]] = []
	anchors: list[int] = []

	entries.sort(key=lambda entry: (entry[0], entry[1]))
	for y, group in groupby(entries, key=lambda entry: entry[0]):
		if bpm is not None:
			time_us += _MEASURE_US * (y - prev_y) / bpm
		time_us += pending_stop
		pending_stop = 0.0

		new_bpm, new_beats, scroll = bpm, beats, None
		for _, _priority, kind, value in group:
			if kind == "bpm":
				new_bpm = value
			elif kind == "meter":
				new_beats = value
			elif kind == "stop":
				pending_stop = _MEASURE_US * value / new_bpm
			elif kind == "scroll":
				scroll = value
			elif kind == "anchor":
				anchors.append(round_ms(time_us / 1000.0))
			else:
				timed[value] = time_us

		if (new_bpm, new_beats) != (bpm, beats):
			bpm_points.append((round_ms(time_us / 1000.0), new_bpm, new_beats, 4))
		if scroll is not None:
			sv_points.append((round_ms(time_us / 1000.0), scroll))
		bpm, beats = new_bpm, new_beats
		prev_y = y

	notes = _build_notes(chart, mode, objects, timed)
	first_note_time = min(
		(time for time, _lane, note_type in notes if note_type is not NoteType.LE),
		default=0,
	)

	timing = build_timing(
		bpm_points=bpm_points,
		sv_points=sv_points,
		first_note_time=first_note_time,
		source=".bms",
		measure_anchor_ms=next((time for time in anchors if time >= first_note_time), None),
	)

	urc_notes = sorted(
		(
			Note(timestamp_ms=time - first_note_time, lane=lane, type=note_type)
			for time, lane, note_type in notes
		),
		key=lambda note: (note.timestamp_ms, note.lane, _TYPE_ORDER[note.type]),
	)
	check_hold_overlap(urc_notes)

	keys, special_keys, special_lanes = _LAYOUTS[mode]
	return Chart(
		format_version=Version(major=1, minor=1),
		metadata=Metadata(
			original="PMS" if chart.pms else "BMS",
			title=chart.title or "Unknown",
			artist=chart.artist or "Unknown",
			creator="Unknown",
			version=chart.play_level or "Unknown",
		),
		judgment=None,
		layout=Layout(keys=keys, special_keys=special_keys, special_lanes=special_lanes),
		timing=timing,
		notes=urc_notes,
	)


def _build_notes(
	chart: BmsChart,
	mode: str,
	objects: list[tuple[float, str, str]],
	timed: list[float],
) -> list[tuple[int, int, NoteType]]:
	streams: dict[str, list[tuple[float, str]]] = {}
	for index, (_y, channel, obj) in enumerate(objects):
		streams.setdefault(channel, []).append((timed[index], obj))

	notes: list[tuple[int, int, NoteType]] = []
	for channel, stream in streams.items():
		lane = _lane(mode, channel)
		if lane is None:
			continue
		kind = _channel_kind(channel)

		if kind == "mine":
			notes.extend((round_ms(time / 1000.0), lane, NoteType.M) for time, _obj in stream)
		elif kind == "ln":
			_pair_long_notes(chart, stream, lane, notes)
		else:
			pending: float | None = None
			for time, obj in stream:
				if chart.lnobj is not None and obj == chart.lnobj and pending is not None:
					notes.append((round_ms(pending / 1000.0), lane, NoteType.LS))
					notes.append((round_ms(time / 1000.0), lane, NoteType.LE))
					pending = None
				else:
					if pending is not None:
						notes.append((round_ms(pending / 1000.0), lane, NoteType.N))
					pending = time
			if pending is not None:
				notes.append((round_ms(pending / 1000.0), lane, NoteType.N))
	return notes


def _pair_long_notes(
	chart: BmsChart,
	stream: list[tuple[float, str]],
	lane: int,
	notes: list[tuple[int, int, NoteType]],
) -> None:
	start: float | None = None
	if chart.lntype == 1:
		for time, obj in stream:
			if obj == "00":
				continue
			if start is None:
				start = time
			else:
				notes.append((round_ms(start / 1000.0), lane, NoteType.LS))
				notes.append((round_ms(time / 1000.0), lane, NoteType.LE))
				start = None
	else:
		for time, obj in stream:
			if obj == "00":
				if start is not None:
					notes.append((round_ms(start / 1000.0), lane, NoteType.LS))
					notes.append((round_ms(time / 1000.0), lane, NoteType.LE))
					start = None
			elif start is None:
				start = time
	if start is not None:
		raise UrcError("syntax", 1, f"long note on lane {lane} has no end")


def _channel_kind(channel: str) -> str | None:
	if len(channel) != 2:
		return None
	first, second = channel
	if first in ("1", "2") and "1" <= second <= "9":
		return "visible"
	if first in ("5", "6") and "1" <= second <= "9":
		return "ln"
	if first in ("D", "E") and "1" <= second <= "9":
		return "mine"
	return None


def _detect_mode(pms: bool, used: set[tuple[int, str]]) -> str:
	if pms:
		if any(
			second in ("6", "7", "8", "9") or (side == 1 and second == "1")
			for side, second in used
		):
			return "PMS18"
		return "PMS9"
	seven = any(second in ("8", "9") for _side, second in used)
	double = any(side == 1 for side, _second in used)
	if seven and double:
		return "14K"
	if double:
		return "10K"
	if seven:
		return "7K"
	return "5K"


def _lane(mode: str, channel: str) -> int | None:
	side = _SIDE_OF[channel[0]]
	key = channel[1]
	if mode in ("5K", "10K"):
		if key == "6":
			return side * 6
		if "1" <= key <= "5":
			return int(key) + side * 6
		return None
	if mode in ("7K", "14K"):
		if key == "6":
			return side * 8
		if "1" <= key <= "5":
			return int(key) + side * 8
		if key in ("8", "9"):
			return int(key) - 8 + 6 + side * 8
		return None
	if mode == "PMS9":
		if side == 0 and "1" <= key <= "5":
			return int(key) - 1
		if side == 1 and "2" <= key <= "5":
			return int(key) - 2 + 5
		return None
	if mode == "PMS18":
		base = side * 9
		if "1" <= key <= "5":
			return base + int(key) - 1
		match key:
			case "8":
				return base + 5
			case "9":
				return base + 6
			case "6":
				return base + 7
			case "7":
				return base + 8
		return None
	return None


def _id_value(text: str, base: int) -> int:
	def digit(char: str) -> int:
		if char.isdigit():
			return int(char)
		if "A" <= char <= "Z":
			return ord(char) - ord("A") + 10
		return ord(char) - ord("a") + 36

	return digit(text[0]) * base + digit(text[1])
